### PR TITLE
Add graphql-doc package recipe

### PR DESCRIPTION
**Pull request by ifitzpatrick**
### Brief summary of what the package does

Uses the GraphQL introspection API to provide clickable documentation inside Emacs for GraphQL endpoints.

### Direct link to the package repository

https://github.com/ifitzpatrick/graphql-doc.el

### Your association with the package

Author

### Relevant communications with the upstream package maintainer

None needed

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

File: recipes/graphql-doc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+(graphql-doc
+	:fetcher github
+	:repo "ifitzpatrick/graphql-doc.el"
+	:files ("graphql-doc.el"))
```
Comment (riscy) [inline recipes/graphql-doc:4]: Prefer the default recipe if possible -- this appears to be equivalent: `(graphql-doc :repo "ifitzpatrick/graphql-doc.el" :fetcher github)`

**Review by riscy — COMMENTED** [review]
Thanks for this, and sorry about the wait -- I've gotten a bit backlogged.  Some quick feedback:

- graphql-doc.el#L26: Prefer `;; Version` over `;; Package-Version` (MELPA automatically adds `Package-Version`)
- graphql-doc.el#L384: Top-level `define-key` can overwrite user bindings.  Try: `(defvar my-map (let ((km (make-sparse-keymap))) (define-key ...) km))`
- In your file's boilerplate, it's a little unusual to split your `URL` and `Author` (and Created, etc.) lines.  Typically they're gathered all up at the top together, if you look at other packages.  I don't think this is a blocker, but something you may want to align with (I think it's easier to read probably because I'm used to it)

Other than those I think this looks good.

**Comment by ifitzpatrick** [comment]
Thank you for the review! I updated the recipe and the package repository with your suggestions. I am new to Emacs lisp so I am a bit confused, why does having `define-key` happen at the top level affect whether it overwrites a users own keybindings?

**Comment by riscy** [comment]
The reason to avoid calling `define-key` at the top level is that users may customize that keymap before the package is loaded (or reloaded) -- for example, imagine if someone wanted to gather all of their keybindings together at the top of their config file -- and then when they get around to loading your package, that work will silently be undone.  Requiring things happen in some "correct" order can force extra complexity onto the user which is why I whine about that one. :)

Taking a closer look now --

A couple minor things I notice in your boilerplate: your URL needs a `.el` on the end, and rather than writing `License: GPLv3` you could look into the one-liner SPDX license identifiers which do the same in a standardized way.  None of those are blockers.

You refer to `:group 'graphql-doc` but you haven't used `defgroup` to create this group yet.  This won't cause an error (the customizaiton interface will just say the definition is missing) but it's good for completeness.

Rather than `(setq inhibit-read-only t) ... (setq inhibit-read-only nil)` you typically would scope that with `(let ((inhibit-read-only t) ...)`, which I think is safer in case something fails.  Ultimately I think that's a minor issue here too.

This looks like a great package and I think the above are minor, so let's go ahead and merge.  But of course feel free to followup if you have any other questions. :)

**Comment by riscy** [comment]
Maybe the more important point about `define-key` is that `defvar` won't set values if they have been set already.

**Comment by ifitzpatrick** [comment]
@riscy Thanks, updated! I have a question about the let binding though; how does `(let ((inhibit-read-only t) ...)` work if I am using lexical binding?

**Comment by riscy** [comment]
Great. :)

I don't think the let binding will change the semantics you're expecting, but that question made me dig into why that's the case.  I believe it boils down to `inhibit-read-only` being a special variable (see [here](https://opensource.com/article/20/3/variables-emacs)).  E.g. with lexical-binding set to t:

```elisp
(setq lexical-binding t)

(let ((inhibit-read-only nil))
  (let ((callback
         (lambda ()
           (format "inhibit-read-only: %s/special: %s" inhibit-read-only
                   (special-variable-p 'inhibit-read-only)))))
    (let ((inhibit-read-only t)) (funcall callback))))
;; => "inhibit-read-only: t/special: t"


(let ((test-var nil))
  (let ((callback
         (lambda ()
           (format "test-var: %s/special: %s" test-var
                   (special-variable-p 'test-var)))))
    (let ((test-var t)) (funcall callback))))
;; => "test-var: nil/special: nil"
```

**Comment by ifitzpatrick** [comment]
Interesting, I did not know about this special variables concept. This is a pretty cool language feature :)

Thanks again for all your help!

**Pull request merged**